### PR TITLE
Fixes to get DCASE evaluation working

### PR DESCRIPTION
**Pull request by turian**
`task_predict.py` is really long and mixes a lot of different logic together. I'm wondering the most logical way to break it down. Perhaps to move `get_events_for_all_files` and its child to another file? I'm not sure honestly.

### CODE REVIEW

File: heareval/predictions/task_predictions.py [hunk 1]
```diff
@@ -91,22 +97,74 @@ def forward(self, x):
...
-        self.log("val_loss", loss, prog_bar=True)
+        return {
+            "predictions": y_pr,
+            "predictions_logit": y_hat,
+            "targets": y,
+            "filenames": filenames,
+            "timestamps": timestamps,
+        }
+
+    def validation_epoch_end(self, outputs):
```
Comment (jorshi) [inline heareval/predictions/task_predictions.py:121]: Cool, good idea to add the validation step here. Instead of trying to fit this all into a single predictor class -- let's have a EventPredictionModel and a ScenePredictionModel?

File: heareval/predictions/task_predictions.py [hunk 2]
```diff
@@ -215,9 +307,33 @@ def create_events_from_prediction(
...
     predictions = (predictions > threshold).astype(np.int8)
+    # print("predictions", predictions)
 
+    event_tree = IntervalTree()
+    for label in range(predictions.shape[1]):
+        # print(label, predictions[:, label])
+        # print(np.where(predictions[:, label]))
+        for group in more_itertools.consecutive_groups(
+            np.where(predictions[:, label])[0]
+        ):
```
Comment (jorshi) [inline heareval/predictions/task_predictions.py:318]: Nice -- this looks clean, wasn't sure how consecutive groups worked -- so this looks for consecutive groups of ones?

File: heareval/predictions/task_predictions.py [hunk 3]
```diff
@@ -146,32 +204,62 @@ def __init__(
         self.labels = pickle.load(
             open(embedding_path.joinpath(f"{split_name}.target-labels.pkl"), "rb")
         )
+        # Only used for event-based prediction
```
Comment (jorshi) [inline heareval/predictions/task_predictions.py:207]: Same here -- different classes of DatasetLoaders for either event or prediction types?

Comment (jorshi) [reply heareval/predictions/task_predictions.py:207]: Then we can avoid all the logic in these functions that is looking for either event or scene.

File: heareval/predictions/task_predictions.py (hunk 2 above)
Comment (turian) [reply heareval/predictions/task_predictions.py:318]: yes

File: heareval/predictions/task_predictions.py (hunk 1 above)
Comment (turian) [reply heareval/predictions/task_predictions.py:122]: Done

File: heareval/predictions/task_predictions.py (hunk 3 above)
Comment (turian) [reply heareval/predictions/task_predictions.py:287]: I think I was able to make this clean by having a dataset metadata dict. It was very simple to change the code to use this. The only downside really is a bit less type-safety and things are a little less explicit.

File: heareval/predictions/task_predictions.py [hunk 4]
```diff
@@ -91,28 +97,160 @@ def forward(self, x):
...
+        z = {
+            "prediction": y_pr,
+            "prediction_logit": y_hat,
+            "target": y,
+        }
+        # https://stackoverflow.com/questions/38987/how-do-i-merge-two-dictionaries-in-a-single-expression-taking-union-of-dictiona
+        return {**z, **metadata}
+
+    # Implement this for each inheriting class
+    # TODO: Can we combine the boilerplate for both of these?
```
Comment (turian) [inline heareval/predictions/task_predictions.py:121]: I'm not sure if we can simplify the boilerplate in validation_epoch_end implementations further

File: heareval/predictions/task_predictions.py [hunk 5]
```diff
@@ -146,38 +286,64 @@ def __init__(
...
             torch.nn.functional.one_hot(torch.LongTensor(y), num_classes=self.nlabels)
             .max(axis=0)
-            .values,
+            .values.float(),
+            self.metadata[idx],
         )
 
 
 def create_events_from_prediction(
     prediction_dict: Dict[float, torch.Tensor],
```
Comment (turian) [inline heareval/predictions/task_predictions.py:341]: Yeah so this params shouldn't be hardcoded in predict really, it should be a part of the event scoring classes. I am not really sure how to move these hyperparams to make them in scores.py but that's where they belong.

Also these methods are really kind confusing here, I think they belong somewhere else.

File: heareval/predictions/task_predictions.py [hunk 6]
```diff
@@ -15,10 +15,13 @@
 """
```
Comment (turian) [inline heareval/predictions/task_predictions.py:15]: This file is too long, but I don't want to break it into ten files just for the sake of it. I wonder what is the most logical way to break it down.

File: heareval/predictions/task_predictions.py [hunk 7]
```diff
@@ -216,44 +386,32 @@ def create_events_from_prediction(
...
+
+            start = timestamps[startidx]
+            end = timestamps[endidx]
+            # Add event if greater than the minimum duration threshold
+            if end - start >= min_duration:
+                # We probably don't need this interval tree and can remove it maybe
+                # from requirements
+                event_tree.addi(begin=start, end=end, data=label)
+
+    # Er why return an intervaltree when we immediately postprocess it to a dict?
```
Comment (turian) [inline heareval/predictions/task_predictions.py:404]: I still want to remove this intervaltree nonsense, it doesn't add anything as far as I can tell?
I don't think we need it at all

File: heareval/predictions/task_predictions.py [hunk 8]
```diff
@@ -349,31 +512,62 @@ def task_predictions_train(
...
 
-    checkpoint_callback = ModelCheckpoint(monitor=target_score, mode="max")
+    if scores[0].maximize:
+        mode = "max"
+    else:
+        mode = "min"
+    checkpoint_callback = ModelCheckpoint(monitor=target_score, mode=mode)
     early_stop_callback = EarlyStopping(
-        monitor=target_score, min_delta=0.00, patience=10, verbose=False, mode="max"
+        # TODO: Tune these
```
Comment (turian) [inline heareval/predictions/task_predictions.py:547]: Still TODO In upcoming PR

File: heareval/predictions/task_predictions.py [hunk 9]
```diff
@@ -349,31 +512,62 @@ def task_predictions_train(
...
     )
     valid_dataloader = dataloader_from_split_name(
-        "valid", embedding_path, label_to_idx, nlabels
+        "valid", embedding_path, label_to_idx, nlabels, metadata["embedding_type"]
     )
     trainer.fit(predictor, train_dataloader, valid_dataloader)
     return predictor
 
 
+# This all needs to be cleaned up and simplified later.
```
Comment (turian) [inline heareval/predictions/task_predictions.py:568]: Still todo in upcoming PR

File: heareval/score.py [hunk 10]
```diff
@@ -171,19 +191,34 @@ class EventBasedScore(SoundEventScore):
...
 
 available_scores: Dict[str, Callable] = {
-    "top1_err": Top1Error,
-    "pitch_err": partial(Top1Error, name="pitch_err"),
-    "chroma_err": ChromaError,
-    "onset_only_event_based": partial(
-        EventBasedScore, params={"evaluate_offset": False, "t_collar": 0.2}
+    "top1_acc": Top1Accuracy,
+    "pitch_acc": partial(Top1Accuracy, name="pitch_acc"),
+    "chroma_acc": ChromaAccuracy,
```
Comment (turian) [inline heareval/score.py:204]: The event glomming hyperparams belong here too somehow.

The score names are getting pretty long but at least they're specific

Comment (jorshi) [reply heareval/score.py:204]: Ya cool -- this is sort of how I imagined this would go, I had thought about moving these score params into the task config dict where we specify the eval metric -- but I think this is okay here.

File: heareval/predictions/task_predictions.py (hunk 5 above)
Comment (jorshi) [reply heareval/predictions/task_predictions.py:341]: Could we move these into the score function somehow? Like a sed eval function that receives a set of frame embedding predictions / targets and then creates the events from those and calculates the scores? So we don't need to fuss with the conversion beforehand?

File: heareval/predictions/task_predictions.py (hunk 7 above)
Comment (jorshi) [reply heareval/predictions/task_predictions.py:404]: One of the remaining things I think we get from intervaltree is the sort of all events by the start timestamp -- so that is kind of nice. I'm curious whether that is necassary for the sed eval stuff. Either way, that wouldn't be too challenging to implement if it is requirement and we have everything we need it looks like to go straight from the consecutive groups to a dictionary w/o the intervaltree

Comment (jorshi) [reply heareval/predictions/task_predictions.py:404]: So if you wanna ditch the interval tree I'd say go for it

File: heareval/predictions/task_predictions.py (hunk 6 above)
Comment (jorshi) [reply heareval/predictions/task_predictions.py:15]: Ya it's a bit long -- but not the worst. I was going to say we could move the prediction models to there own module -- but maybe it is kind of nice to have them in the same file. I don't think it is a huge issue. I'm kind of into the idea of moving the event creation from frames functions into the scoring modules.

File: heareval/predictions/task_predictions.py (hunk 3 above)
Comment (jorshi) [reply heareval/predictions/task_predictions.py:288]: hmm ya. interesting -- on one hand it creates some flexible down the road if we want to pass other types of metadata depending on the task type. It's not ideal have to pass around the empty dict for the scene stuff. I can't think of a better approach atm though, this seems to be working pretty well.

**Review by jorshi — COMMENTED** [review]
Took an initial pass on this -- overall awesome work, I think this is looking good. Made a few comments.

File: heareval/predictions/task_predictions.py (hunk 2 above)
Comment (turian) [reply heareval/predictions/task_predictions.py:392]: The previous implementation had a bug on the boundaries

File: heareval/predictions/task_predictions.py (hunk 3 above)
Comment (khumairraj) [reply heareval/predictions/task_predictions.py:288]: Yes. The metadata dict looks nice and it does give more flexibility.

File: heareval/predictions/task_predictions.py (hunk 5 above)
Comment (khumairraj) [reply heareval/predictions/task_predictions.py:341]: We might want to use this to dump the events in the test part (I see commented code for the test). But putting in score also sounds like a good option

**Review by khumairraj — APPROVED** [review]
I think the abstraction are great and the event-based task are fitting in well.
